backend/internal/database/book.go

```diff
@@ -297,9 +297,129 @@ func (db *BookDatabase) IncrementBookCounter(ctx context.Context, id string) err
  */
 func (db *BookDatabase) FetchBookAnalytics(ctx context.Context, id string, dRange int) ([]int, error) {
 	var analytics []int
+
+	var firstYearStart int
+	var firstYearEnd int
+	var secondYearEnd int
+
+	firstYearStart, firstYearEnd, secondYearEnd = AnalyticsClicksIndices(dRange)
+
+	var query string = "SELECT ARRAY_CAT(clicks[$1:$2], clicks[1:$3]) " +
+		"FROM book_analytics WHERE id = $4"
+	rows, err := db.Conn.Query(ctx, query, firstYearStart, firstYearEnd, secondYearEnd, id)
+
+	if err != nil {
+		fmt.Print(err)
+		return analytics, errors.Wrap(err, "error on query for book analytics")
+	}
+
+	defer rows.Close()
+
+	if !rows.Next() {
+		return analytics, nil
+	}
+
+	err = rows.Scan(&analytics)
+
+	if err != nil {
+		return analytics, errors.Wrap(err, "error on Scan for book clicks array")
+	}
+
+	return analytics, err
+
+}
+
+/*
+ * Returns a dictionary mapping each book's id to its daily click counts given a range from the current day
+ */
+func (db *BookDatabase) FetchAllBookAnalytics(ctx context.Context, dRange int) ([]models.Analytic, error) {
+	analytics := make([]models.Analytic, 0)
+
+	var firstYearStart int
+	var firstYearEnd int
+	var secondYearEnd int
+
+	firstYearStart, firstYearEnd, secondYearEnd = AnalyticsClicksIndices(dRange)
+
+	var query string = "SELECT id, ARRAY_CAT(clicks[$1:$2], clicks[1:$3]) " +
+		"FROM book_analytics"
+	rows, err := db.Conn.Query(ctx, query, firstYearStart, firstYearEnd, secondYearEnd)
+
+	if err != nil {
+		fmt.Print(err)
+		return analytics, errors.Wrap(err, "error on query for all book analytics")
+	}
+
+	defer rows.Close()
+
+	for rows.Next() {
+		var analytic models.Analytic
+		if err := rows.Scan(&analytic.ID, &analytic.Clicks); err != nil {
+			fmt.Print(err)
+			return analytics, errors.Wrap(err, "error on Scan for book id and clicks array in "+
+				"FetchAllBookAnalytics")
+		}
+
+		analytics = append(analytics, analytic)
+	}
+
+	return analytics, err
+
+}
+
+/*
+ * Gets 5 most popular books from the past 30 days based on click analytics
+ */
+func (db *BookDatabase) FetchPopularBooks(ctx context.Context) ([]models.Book, error) {
```

Comment:
This function should use the book_analytics_last_30 view IMO, basically select the top 5 from that table (using the `total` column) and join it with `books`.